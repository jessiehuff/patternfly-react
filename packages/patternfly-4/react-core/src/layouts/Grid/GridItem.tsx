import * as React from 'react';
import styles from '@patternfly/react-styles/css/layouts/Grid/grid';
import { css } from '@patternfly/react-styles';
import { DeviceSizes } from '../../styles/sizes';
import {
  getOffsetKey,
  getOffsetModifier,
  getRowSpanKey,
  getRowSpanModifier,
  getSpanModifier,
  gridSpans
} from './gridUtils';

export type gridSpans = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export interface GridItemProps extends React.HTMLProps<HTMLDivElement> {
   /** content rendered inside the Grid Layout Item */
   children?: React.ReactNode;
   /** additional classes added to the Grid Layout Item */
   className?: string;
   /** the number of columns the grid item spans. Value should be a number 1-12   */
   span?: gridSpans;
   /** the number of rows the grid item spans. Value should be a number 1-12   */
   rowSpan?: gridSpans;
   /** the number of columns a grid item is offset */
   offset?: gridSpans;
   /** the number of columns the grid item spans on small device. Value should be a number 1-12   */
   sm?: gridSpans;
   /** the number of rows the grid item spans on medium device. Value should be a number 1-12   */
   smRowSpan?: gridSpans;
   /** the number of columns the grid item is offset on small device. Value should be a number 1-12   */
   smOffset?: gridSpans;
   /** the number of columns the grid item spans on medium device. Value should be a number 1-12   */
   md?: gridSpans;
   /** the number of rows the grid item spans on medium device. Value should be a number 1-12   */
   mdRowSpan?: gridSpans;
   /** the number of columns the grid item is offset on medium device. Value should be a number 1-12   */
   mdOffset?: gridSpans;
   /** the number of columns the grid item spans on large device. Value should be a number 1-12   */
   lg?: gridSpans;
   /** the number of rows the grid item spans on large device. Value should be a number 1-12   */
   lgRowSpan?: gridSpans;
   /** the number of columns the grid item is offset on large device. Value should be a number 1-12   */
   lgOffset?: gridSpans;
   /** the number of columns the grid item spans on xLarge device. Value should be a number 1-12   */
   xl?: gridSpans;
   /** the number of rows the grid item spans on large device. Value should be a number 1-12   */
   xlRowSpan?: gridSpans;
   /** the number of columns the grid item is offset on xLarge device. Value should be a number 1-12   */
   xlOffset?: gridSpans;
   /** the number of columns the grid item spans on 2xLarge device. Value should be a number 1-12   */
   xl2?: gridSpans;
   /** the number of rows the grid item spans on 2xLarge device. Value should be a number 1-12   */
   xl2RowSpan?: gridSpans;
   /** the number of columns the grid item is offset on 2xLarge device. Value should be a number 1-12   */
   xl2Offset?: gridSpans;
}
export const GridItem: React.FunctionComponent<GridItemProps> = ({
  children = null,
  className = '',
  span = null,
  rowSpan = null,
  offset = null,
  sm = null,
  smRowSpan = null,
  smOffset = null,
  md = null,
  mdRowSpan = null,
  mdOffset = null,
  lg = null,
  lgRowSpan = null,
  lgOffset = null,
  xl = null,
  xlRowSpan = null,
  xlOffset = null,
  xl2 = null,
  xl2RowSpan = null,
  xl2Offset = null, 
  ...props
}: GridItemProps) => {
  const classes = [
    styles.gridItem,
    span && getSpanModifier(span),
    offset && getOffsetModifier(offset),
    rowSpan && getRowSpanModifier(rowSpan)
  ];

  Object.entries(DeviceSizes).forEach(([propKey, classModifier]) => {
    const spanValue = props[propKey];
    const rowSpanValue = props[getRowSpanKey(propKey)];
    const offsetValue = props[getOffsetKey(propKey)];

    if (spanValue) {
      classes.push(getSpanModifier(spanValue, classModifier));
    }
    if (rowSpanValue) {
      classes.push(getRowSpanModifier(rowSpanValue, classModifier));
    }
    if (offsetValue) {
      classes.push(getOffsetModifier(offsetValue, classModifier));
    }

    delete props[propKey];
    delete props[getRowSpanKey(propKey)];
    delete props[getOffsetKey(propKey)];
  });

  return (
    <div className={css(...classes, className)} {...props}>
      {children}
    </div>
  );
};