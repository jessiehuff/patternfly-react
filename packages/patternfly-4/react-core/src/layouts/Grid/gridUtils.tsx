import styles from '@patternfly/react-styles/css/layouts/Grid/grid';
import { getModifier } from '@patternfly/react-styles';
import { OneOf } from '../../helpers/typeUtils';
import { DeviceSizes } from '../../styles/sizes';

export const gridSpans = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

export const getRowSpanKey = (size: string) => `${size}RowSpan` as string;

export const getOffsetKey = (size: string) => `${size}Offset` as string;

export const getSizeSuffix = (size: string) => (size ? `-on-${size}` : '') as string;

export const getSpanModifier = (value: number, size: 'sm' | 'md' | 'lg' | 'xl' | 'xl2') => getModifier(styles, `${value}-col${getSizeSuffix(size)}`);

export const getGridSpanModifier = (value: number, size: 'sm' | 'md' | 'lg' | 'xl' | 'xl2') => getModifier(styles, `-all-${value}-col${getSizeSuffix(size)}`);

export const getRowSpanModifier = (value: number, size: 'sm' | 'md' | 'lg' | 'xl' | 'xl2') => getModifier(styles, `${value}-row${getSizeSuffix(size)}`);

export const getOffsetModifier = (value: number, size: 'sm' | 'md' | 'lg' | 'xl' | 'xl2') => getModifier(styles, `offset-${value}-col${getSizeSuffix(size)}`);
