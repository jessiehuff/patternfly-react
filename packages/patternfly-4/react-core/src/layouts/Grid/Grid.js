import React from 'react';
import PropTypes from 'prop-types';
import styles from '@patternfly/react-styles/css/layouts/Grid/grid';
import { css } from '@patternfly/react-styles';
import { GutterSize, getGutterModifier } from '../../styles/gutters';
import { getGridSpanModifier, gridSpans } from './gridUtils';
import { DeviceSizes } from '../../styles/sizes';

const gridItemSpanValueShape = PropTypes.oneOf(gridSpans);

const propTypes = {
  /** content rendered inside the Grid layout */
  children: PropTypes.any,
  /** additional classes added to the Grid layout */
  className: PropTypes.string,
  /** Adds space between children. Options are sm, md or lg */
  gutter: PropTypes.oneOf(Object.keys(GutterSize)),
  /** The number of rows a column in the grid should span.  Value should be a number 1-12 */
  span: gridItemSpanValueShape,
  /** the number of columns all grid items should span on a small device */
  sm: gridItemSpanValueShape,
  /** the number of columns all grid items should span on a medium device */
  md: gridItemSpanValueShape,
  /** the number of columns all grid items should span on a large device */
  lg: gridItemSpanValueShape,
  /** the number of columns all grid items should span on a xLarge device */
  xl: gridItemSpanValueShape,
  /** the number of columns all grid items should span on a 2xLarge device */
  xl2: gridItemSpanValueShape,
  /** Additional props are spread to the container <div> */
  '': PropTypes.any // eslint-disable-line react/require-default-props
};

const defaultProps = {
  children: null,
  className: '',
  gutter: null,
  span: null,
  sm: null,
  md: null,
  lg: null,
  xl: null,
  xl2: null
};

const Grid = ({ children, className, gutter, span, ...props }) => {
  const classes = [styles.grid, span && getGridSpanModifier(span)];

  Object.entries(DeviceSizes).forEach(([propKey, gridSpanModifier]) => {
    const propValue = props[propKey];
    if (propValue) {
      classes.push(getGridSpanModifier(propValue, gridSpanModifier));
    }
    delete props[propKey];
  });

  return (
    <div
      className={css(
        ...classes,
        gutter && getGutterModifier(styles, gutter, styles.modifiers && styles.modifiers.gutter),
        className
      )}
      {...props}
    >
      {children}
    </div>
  );
};

Grid.propTypes = propTypes;
Grid.defaultProps = defaultProps;

export default Grid;
