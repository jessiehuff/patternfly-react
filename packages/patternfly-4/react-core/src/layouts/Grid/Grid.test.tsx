import * as React from 'react';
import { Grid } from './Grid';
import { shallow } from 'enzyme';

[true, false].forEach(hasGutter => {
  test(`hasGutter ${hasGutter}`, () => {
    const view = shallow(<Grid hasGutter />);
    expect(view).toMatchSnapshot();
  });
});
