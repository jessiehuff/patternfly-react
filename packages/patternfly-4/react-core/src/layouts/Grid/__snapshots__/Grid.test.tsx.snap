// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hasGutter false 1`] = `
<div
  className="pf-l-grid pf-m-gutter"
/>
`;

exports[`hasGutter true 1`] = `
<div
  className="pf-l-grid pf-m-gutter"
/>
`;
